src/index.js

```diff
@@ -185,6 +185,43 @@ function getMatchingElements (elementIterator, ast, multiple) {
   return results
 }
 
+function getMatchingElementsByTagName (elementIterator, tagName) {
+  const results = [];
+  let element
+  while ((element = elementIterator.next())) {
+    if(tagName.toLowerCase() === element.tagName.toLowerCase() || tagName === '*') {
+      results.push(element);
+    }
+  }
+  return results
+}
+
+/**
+ * https://dom.spec.whatwg.org/#concept-getelementsbytagnamens
+ */
+function getMatchingElementsByTagNameNS (elementIterator, namespaceURI, tagName) {
+  const results = [];
+  // exit early if null, empty string or undefined is provided
+  // these will not match the element namespace
+  if (!namespaceURI) {
+    return results;
+  }
+  let element
+  
+  while ((element = elementIterator.next())) {
+
+    const match = element.outerHTML.match(/^<\s*([\w-]+)/);
```

Comment:
can't match against tagName because it's always uppercase

Reply:
the `\s*` isn't valid XHTML I think but does it hurt... not sure how likely we'll get invalid HTML?

Reply:
This is strange to me. So to match browser behavior for case-sensitivity, we seriously have to get the `tagName` this way? The regex makes me nervous, but I'm struggling to think of a better way...

What if instead, we did this:

1. Get `element.outerHTML`
2. Convert it to uppercase
3. Find `outerHtmlAsUppercase.indexOf(element.tagName)`
4. Use that index to get the substring in the original `element.outerHTML`.

This seems more elaborate to me, but may be a bit less fragile than a regex... Not sure, what do you think?

Reply:
Yeah I'm not a fan of having to do this. Especially since it's super rare to actually care about case sensitivity.

But the `indexOf` approach is going to have false positives, what if I have `<div class="span">`? The regex is solid and I think we can guarantee the tagName is always preceded by a `<` and followed by a space?

I tried to add more tests to give us better confidence in the regex with wacky HTML but each time I ended up with invalid HTML and not sure it makes sense to test against invalid HTML.

Reply:
In the case of `<div class="span">`, wouldn't the `tagName` be `DIV`, not `SPAN`?

I guess mostly I'm worried that something about `\s*` or `[\w-]+` may come back to bite us. What if there's a newline – that doesn't count as `\s` unless we also add `/s` to the end of the regex, which is only supported in recent browsers. Does `[\w-]+` really match all possible HTML tag names? Underscores should be allowed here as well. I just tested, and Chrome also allows emoji in tag names (so `<hello-👪>foo</hello-👪>` is valid).

If we use `indexOf()` then we don't have to worry about these kinds of edge cases, although I agree it's no good if it has its own edge cases. 🙂

Reply:
Sure, let me try it and we'll see how it looks